server/src/services/initialization/indexWorkspaceFolders.ts

```diff
@@ -23,79 +29,111 @@ export interface IndexWorkspaceFoldersContext {
 }
 
 export async function indexWorkspaceFolders(
-  indexWorkspaceFoldersContext: IndexWorkspaceFoldersContext,
+  context: IndexWorkspaceFoldersContext,
   workspaceFileRetriever: WorkspaceFileRetriever,
   workspaceFolders: WorkspaceFolder[]
 ) {
-  const { logger } = indexWorkspaceFoldersContext;
+  const { logger } = context;
 
   if (workspaceFolders.some((wf) => wf.uri.includes("\\"))) {
     throw new Error("Unexpect windows style path");
   }
 
-  indexWorkspaceFoldersContext.indexJobCount++;
-  const indexJobId = indexWorkspaceFoldersContext.indexJobCount;
+  context.indexJobCount++;
+  const indexJobId = context.indexJobCount;
 
   const indexJobStartTime = new Date();
   logger.info(`[indexing:${indexJobId}] Starting indexing job ...`);
 
-  notifyStartIndexing(indexJobId, indexWorkspaceFoldersContext);
+  notifyStartIndexing(indexJobId, context);
 
   const topLevelWorkspaceFolders = resolveTopLevelWorkspaceFolders(
-    indexWorkspaceFoldersContext,
+    context,
     workspaceFolders
   );
 
   if (topLevelWorkspaceFolders.length === 0) {
     notifyNoOpIndexing(
       indexJobId,
-      indexWorkspaceFoldersContext,
+      context,
       `[indexing:${indexJobId}] Workspace folders already indexed`
     );
 
     return;
   }
 
   logger.info(`[indexing:${indexJobId}] Workspace folders`);
+
   for (const workspaceFolder of topLevelWorkspaceFolders) {
     logger.info(`[indexing:${indexJobId}]   ${workspaceFolder.name}`);
-  }
 
-  for (const workspaceFolder of topLevelWorkspaceFolders) {
-    try {
-      await scanForHardhatProjectsAndAppend(
-        indexJobId,
-        workspaceFolder,
-        indexWorkspaceFoldersContext.projects,
-        workspaceFileRetriever,
-        logger
-      );
-    } catch (err) {
-      logger.error(err);
+    const hardhatConfigFiles = await findHardhatConfigFiles(
+      workspaceFolder,
+      workspaceFileRetriever
+    );
+
+    const foundryConfigFiles = await findFoundryConfigFiles(
+      workspaceFolder,
+      workspaceFileRetriever
+    );
+
+    const hardhatProjects = await Promise.all(
+      hardhatConfigFiles.map((configFile) =>
+        buildProject<HardhatProject>(
+          HardhatProject,
+          configFile,
+          workspaceFolder,
+          workspaceFileRetriever
+        )
+      )
+    );
+
+    const foundryProjects = await Promise.all(
+      foundryConfigFiles.map((configFile) =>
+        buildProject<FoundryProject>(
+          FoundryProject,
+          configFile,
+          workspaceFolder,
+          workspaceFileRetriever
+        )
+      )
+    );
+
+    for (const project of [...hardhatProjects, ...foundryProjects]) {
```

Comment:
I am trying to remember, this is a guard against projects nested in projects?

Does this need to consider foundry in hh or hh within foundry? Or are we sticking to a rule that says, only the top level project counts?

Reply:
Lets deal with that when we do the refactor.